openhands/runtime/impl/action_execution/action_execution_client.py

```diff
@@ -358,50 +361,55 @@ def get_updated_mcp_config(
     ) -> MCPConfig:
         # Add the runtime as another MCP server
         updated_mcp_config = self.config.mcp.model_copy()
-        
+
         # Get current stdio servers
-        current_stdio_servers: list[MCPStdioServerConfig] = list(updated_mcp_config.stdio_servers)
+        current_stdio_servers: list[MCPStdioServerConfig] = list(
+            updated_mcp_config.stdio_servers
+        )
         if extra_stdio_servers:
             current_stdio_servers.extend(extra_stdio_servers)
-            
-        # Convert to JSON format for comparison
-        current_json_configs = [server.model_dump(mode='json') for server in current_stdio_servers]
-        
-        # Create a lookup of server JSON by its string representation for fast comparison
-        last_updated_json_set = {
-            json.dumps(server.model_dump(mode='json'), sort_keys=True)
-            for server in self._last_updated_mcp_stdio_servers
-        }
-        
-        # Check if there are any new servers
-        new_servers = []
-        for idx, config in enumerate(current_json_configs):
-            if json.dumps(config, sort_keys=True) not in last_updated_json_set:
-                new_servers.append(current_stdio_servers[idx])
-        
-        self.log('debug', f'adding {len(new_servers)} new stdio servers to MCP config: {new_servers}')
-        
+
+        # Check if there are any new servers using the __eq__ operator
+        new_servers = [
+            server
+            for server in current_stdio_servers
+            if server not in self._last_updated_mcp_stdio_servers
+        ]
+
+        self.log(
+            'debug',
+            f'adding {len(new_servers)} new stdio servers to MCP config: {new_servers}',
+        )
+
         # Only send update request if there are new servers
         if new_servers:
             # Use all current servers for the update (simpler and more reliable)
             # This works because we're replacing the complete configuration
-            stdio_tools = [server.model_dump(mode='json') for server in current_stdio_servers]
+            stdio_tools = [
```

Comment:
@OpenHands the `stdio_tools` we send should be a union of `self._last_updated_mcp_stdio_servers` and `current_stdio_servers`